components/idenpub/idenpub.go

```diff
@@ -0,0 +1,127 @@
+package idenpub
+
+import (
+	"bytes"
+
+	"github.com/iden3/go-iden3-core/db"
+	"github.com/iden3/go-iden3-core/merkletree"
+)
+
+var (
+	rotRootKey = []byte("rotRoot")
+	retRootKey = []byte("retRoot")
+	rotKey     = []byte("rot")
+	retKey     = []byte("ret")
+)
+
+// IdenPub is a interface, that for the moment will be satisfied at least by IdenPubHTTP & IdenPubIPFS.
+type IdenPub interface {
+	Publish()
+}
+
+// IdenPubHTTP satisfies the IdenPub interface, and stores in a leveldb the published RoT & ReT to be returned when requested.
+type IdenPubHTTP struct {
+	db  db.Storage
+	rot *merkletree.MerkleTree
+	ret *merkletree.MerkleTree
+}
+
+// AddLeafRoT adds a new leaf to the given MerkleTree, which contains the Root
+func AddLeafRoT(mt *merkletree.MerkleTree, root merkletree.Hash) error {
+	l := NewLeafRoT(root)
+	return mt.AddEntry(l.Entry())
+}
+
+// AddLeafReT adds a new leaf to the given MerkleTree, which contains the Nonce & Version
+func AddLeafReT(mt *merkletree.MerkleTree, nonce, version uint32) error {
+	l := NewLeafReT(nonce, version)
+	return mt.AddEntry(l.Entry())
+}
+
+// NewIdenPubHTTP returns a new IdenPubHTTP
+func NewIdenPubHTTP(db db.Storage, rot *merkletree.MerkleTree, ret *merkletree.MerkleTree) *IdenPubHTTP {
+	return &IdenPubHTTP{
+		db:  db,
+		rot: rot,
+		ret: ret,
+	}
+}
+
+// Publish publishes the RoT and ReT to the configured way of publishing
+func (i *IdenPubHTTP) Publish() error {
```

Comment:
As discussed offline, it will be better for the IdenPub to maintain a copy of the t and t-1 IdenState related trees.  And Instead of just publishing the leafs, publish the entire tree so that clients don't need to spend time calculating all the hashes (which with Poseidon are expensive)